File: src/AAS_Cores/ROS_AAS_Core/src/utilities/AASArchive_utils.py
"""File to group useful methods for accessing and managing the AAS Archive."""
import calendar
import json
import logging
import os
import time

from utilities.AASarchiveInfo import AASarchiveInfo

_logger = logging.getLogger(__name__)


# ------------------------
# Methods related to files
# ------------------------
def create_status_file():
    """This method creates the status file of the AAS Manager and sets it to "initializing"."""
    initial_status_info = {'name': 'AAS_Core', 'status': 'Initializing', 'timestamp': calendar.timegm(time.gmtime())}

    # with (open(AASarchiveInfo.CORE_STATUS_FILE_PATH, 'x') as status_file):
    #     json.dump(initial_status_info, status_file)
    #     status_file.close()
    f = open(AASarchiveInfo.CORE_STATUS_FILE_PATH, 'x')
    json.dump(initial_status_info, f)
    f.close()

def create_interaction_files():
    """This method creates the necessary interaction files to exchange information between AAS Core and AAS Manager."""

    # First interaction folders are created
    os.mkdir(AASarchiveInfo.CORE_INTERACTIONS_FOLDER_PATH)
    os.mkdir(AASarchiveInfo.MANAGER_INTERACTIONS_FOLDER_PATH)

    # Then the interaction files are added in each folder
    core_requests_file = open(AASarchiveInfo.CORE_INTERACTIONS_FOLDER_PATH + AASarchiveInfo.SVC_REQUEST_FILE_SUBPATH,'x')
    core_responses_file = open(AASarchiveInfo.CORE_INTERACTIONS_FOLDER_PATH + AASarchiveInfo.SVC_RESPONSE_FILE_SUBPATH,'x')
    manager_requests_file = open(AASarchiveInfo.MANAGER_INTERACTIONS_FOLDER_PATH + AASarchiveInfo.SVC_REQUEST_FILE_SUBPATH,'x')
    manager_responses_file = open(AASarchiveInfo.MANAGER_INTERACTIONS_FOLDER_PATH + AASarchiveInfo.SVC_RESPONSE_FILE_SUBPATH,'x')

    core_requests_file.write('{"serviceRequests": []}')
    core_requests_file.close()

    manager_requests_file.write('{"serviceRequests": []}')
    manager_requests_file.close()

    core_responses_file.write('{"serviceResponses": []}')
    core_responses_file.close()

    manager_responses_file.write('{"serviceResponses": []}')
    manager_responses_file.close()

def file_to_json(file_path):
    """
    This method gets the content of a JSON file.

    Args:
        file_path (str): the path of the JSON file.

    Returns:
        dict: content of the file in JSON format."""
    f = open(file_path)
    try:
        content = json.load(f)
        f.close()
    except json.JSONDecodeError as e:
        _logger.error("Invalid JSON syntax:" + str(e))
        return None
    return content

def update_json_file(file_path, content):
    """
    This method updates the content of a JSON file.

    Args:
        file_path (str): the path to the JSON file.
        content (dict): the content of the JSON file.
    """
    with open(file_path, "w") as outfile:
        json.dump(content, outfile)

File: src/AAS_Cores/Numerical_AAS_Core/src/main/java/utilities/AAS_ArchiveUtils.java
package utilities;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.*;

public class AAS_ArchiveUtils {


    // ------------------------
    // Methods related to files
    // ------------------------
    public static void createStatusFile() {
        JSONObject statusFileJSON = new JSONObject();
        statusFileJSON.put("name", "AAS_Core");
        statusFileJSON.put("status", "Initializing");
        statusFileJSON.put("timestamp", System.currentTimeMillis() / 1000);

        updateFile(AAS_ArchiveInfo.coreStatusFilePath, statusFileJSON);
    }

    public static void changeStatus(String newStatus) {
        JSONObject statusFileJSON = fileToJSON(AAS_ArchiveInfo.coreStatusFilePath);
        statusFileJSON.put("status", newStatus);
        updateFile(AAS_ArchiveInfo.coreStatusFilePath, statusFileJSON);
    }

    public static String getManagerStatus() {
        JSONParser parser = new JSONParser();
        try (Reader reader = new FileReader(AAS_ArchiveInfo.managerStatusFilePath)) {
            JSONObject statusFileJSON = (JSONObject) parser.parse(reader);
            return statusFileJSON.get("status").toString();
        }  catch (FileNotFoundException e) {
            return null;
        } catch (IOException | ParseException e) {
            throw new RuntimeException(e);
        }
    }

    public static JSONObject fileToJSON(String filePath) {
        JSONParser parser = new JSONParser();
        try (Reader reader = new FileReader(filePath)) {
            return (JSONObject) parser.parse(reader);
        } catch (IOException | ParseException e) {
            throw new RuntimeException(e);
        }
    }

    public static void updateFile(String filePath, JSONObject content) {
        try (FileWriter file = new FileWriter(filePath)) {
            file.write(content.toJSONString());
            file.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    // -----------------------------------
    // Methods related to service requests
    // -----------------------------------
    public static JSONObject getNextSvcRequest() {
        /**
         * This method gets the next service request that has not be realized. To achieve that it analyzes the service
         * requests of AAS Manager, and check if there is an response for each request in the AAS Core, because if
         * there is not, it means that the request has to be performed. Returns the JSON object of the new request or
         * null if no request has been made.
         * @return JSON object with the information of the new service. NULL if it no request has been made.
         */

        // For each service, check if there is an response for each request, because if there is not, it means that the
        // request has to be performed
        JSONArray requestsArray = (JSONArray) fileToJSON(AAS_ArchiveInfo.managerInteractionsFolderPath + AAS_ArchiveInfo.svcRequestFileSubPath).get("serviceRequests");
        for (Object obj: requestsArray) {
            JSONObject reqObj = (JSONObject) obj;
            if (getResponseServiceJSON(((Long) reqObj.get("interactionID")).intValue()) == null)
                return reqObj;
        }
        return null;
    }

    // ------------------------------------
    // Methods related to service responses
    // ------------------------------------
    public static void setResponseServiceState(int interactionID, String newState) {
        // Conseguimos el JSON del servicio en el archivo de respuestas
        JSONObject serviceJSON = getResponseServiceJSON(interactionID);
        if (serviceJSON == null) {
            serviceJSON = new JSONObject();
            serviceJSON.put("interactionID", interactionID);
        }
        serviceJSON.put("serviceStatus", newState);
        updateSvcResponse(serviceJSON);
    }

    // ---------------------------------------
    // Methods related to responses (by AAS Core)
    //----------------------------------------
    private static void updateSvcResponse(JSONObject serviceJSON) {
        JSONArray responsesArray = (JSONArray) fileToJSON(AAS_ArchiveInfo.coreInteractionsFolderPath + AAS_ArchiveInfo.svcResponseFileSubPath).get("serviceResponses");
        responsesArray.add(serviceJSON);
        JSONObject updatedContent = new JSONObject();
        updatedContent.put("serviceResponses", responsesArray);
        updateFile(AAS_ArchiveInfo.coreInteractionsFolderPath + AAS_ArchiveInfo.svcResponseFileSubPath, updatedContent);
    }

    private static JSONObject getResponseServiceJSON(int interactionID) {
        JSONArray responsesArray = (JSONArray) fileToJSON(AAS_ArchiveInfo.coreInteractionsFolderPath + AAS_ArchiveInfo.svcResponseFileSubPath).get("serviceResponses");
        for (Object o: responsesArray) {
            JSONObject svcJSON = (JSONObject) o;
            if (interactionID == ((Long) svcJSON.get("interactionID")).intValue())
                return svcJSON;
        }
        return null;
    }

    public static JSONObject createSvcCompletedResponse(JSONObject requestJSON, String serviceRequestedData) {
        JSONObject completedResponseJSON = new JSONObject();
        completedResponseJSON.put("interactionID", requestJSON.get("interactionID"));
        completedResponseJSON.put("thread", requestJSON.get("thread"));
        completedResponseJSON.put("serviceID", requestJSON.get("serviceID"));
        completedResponseJSON.put("serviceType", requestJSON.get("serviceType"));

        JSONObject serviceDataJSON = new JSONObject();
        serviceDataJSON.put("serviceStatus", "Completed");
        serviceDataJSON.put("serviceCategory", "service-response");
        serviceDataJSON.put("timestamp", System.currentTimeMillis() / 1000);   // Add the timestamp in seconds

        if (serviceRequestedData != null) {
            String requestedDataName = (String) ((JSONObject) ((JSONObject) requestJSON.get("serviceData")).get("serviceParams")).get("requestedData");
            JSONObject requestedDataJSON = new JSONObject();
            requestedDataJSON.put(requestedDataName, serviceRequestedData);
            serviceDataJSON.put("serviceParams", requestedDataJSON);
        }
        completedResponseJSON.put("serviceData", serviceDataJSON);
        return completedResponseJSON;
    }

    public static void updateSvcCompleteResponse(JSONObject responseFinalJSON) {
        String svcResponsesFilePath = AAS_ArchiveInfo.coreInteractionsFolderPath + AAS_ArchiveInfo.svcResponseFileSubPath;
        System.out.println(fileToJSON(svcResponsesFilePath).toJSONString());
        JSONArray responsesArray = (JSONArray) fileToJSON(svcResponsesFilePath).get("serviceResponses");
        if (responsesArray == null)
            responsesArray = new JSONArray();
        responsesArray.add(responseFinalJSON);
        JSONObject updatedResponseContent = new JSONObject();
        updatedResponseContent.put("serviceResponses", responsesArray);
        updateFile(svcResponsesFilePath, updatedResponseContent);
    }
}
